---
title: 'Towards interoptability between SingleCellExperiment-derived packages.'
author: Andrew McDavid
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Interoptability between SingleCellExperiment-derived packages}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

## Install some devel packages, load data.

```{r, results = 'hide'}
## BiocInstaller::biocLite(c("drisso/SingleCellExperiment",
## "RGLab/MAST@SingleCellExperiment",
## "drisso/zinbwave@SCE"))
suppressPackageStartupMessages({library(MAST)
library(zinbwave)
library(magrittr)
library(ggplot2)
library(GGally)
library(dplyr)
library(dtplyr)
})

knitr::opts_chunk$set(cache = TRUE, message = FALSE,error = FALSE,warning = FALSE)
```

Install some development packages.


```{r data,results='hide'}
data(maits, package='MAST')
dim(maits$expressionmat)
unlog <- function(x) round(2^x - 1)
maits$fdata$primerid <- colnames(maits$expressionmat) <- make.names(maits$fdata$symbolid)
sca_raw <- FromMatrix(maits$expressionmat %>% t() %>% unlog() , maits$cdat, maits$fdat)
sca_filt <- sca_raw[freq(sca_raw)>.1,] %>% subset(., ourfilter)
sca_filt <- sca_filt[sample(nrow(sca_filt), 1500),]

```

These data are log2 + 1 transcripts per million (TPM), as output by RSEM--but other packages expect count-like data.

```{r}
zw <- zinbwave(Y = sca_filt, K = 2)

```

Run zinbwave on two factors


```{r}
ggpairs(data.frame(zw = reducedDim(zw), colData(zw)), columns = c('zw.W1', 'zw.W2', 'ngeneson'), mapping = aes(color = condition))

```

## Using MAST to characterizing genes that drive the factors

```{r, results = 'hide'}
assays(zw)[[1]] <- log2(assay(zw)+1)
colData(zw) <- cbind(colData(zw), zw=as.data.frame(reducedDim(zw)))
zw <- zw %>% as('SingleCellAssay')
zz <- zlm(~W1+W2, sca=zw)

```

```{r, results = 'asis'}
ss <- summary(zz)
knitr::kable(print(ss))

```

These are log-fold changes in the top few changes associated with factors 1 and 2.

```{r}
top5 <- ss$datatable %>% filter(component=='logFC', contrast %like% 'W') %>% arrange(-abs(z)) %>% head(n=5) %>% left_join(rowData(zw) %>% as.data.table())
plt <- ggplot(zw[top5$primerid,] %>% as('data.table'), aes(x=W1, color = condition)) + geom_point() + facet_wrap(~symbolid)

```



```{r}
plt + aes(y = et)
```

Expression on "Et" scale ($\log_2( TPM + 1)$)

```{r}
plt + aes(y = normalizedValues)
```

Normalized expression from factor model

## Potential benefits:

*  Improve and speed analyses -- mix and match normalization methods,  dimensionality reduction, plotting
*  Improve statistical methods -- allow inter-method comparison
*  Improve software -- write utility functions once, and well

## Potential pitfalls:

*  What is in slot 1 of assays? Counts? Normalized counts? Log-transformed?
*  Implicit contracts between functions (load -- normalize -- analysis )
*  Idiosyncratic results ( analysis -- plot )

# SingleCellExperiment: ask for it from your local bioinformatic package today! #
