<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using MAST with RNASeq: MAIT Analysis. • MAST</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MAST</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RGLab/MAST/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Using MAST with RNASeq: MAIT Analysis.</h1>
                        <h4 class="author">Greg Finak, Andrew McDavid, Masanao Yajima, Jingyuan Deng, Vivian Gersuk, Alex Shalek, Chloe K. Schlicter, Hannah W. Miller, M. Juliana McElrath, Martin Prlic, Peter Linsley, Raphael Gottardo</h4>
            
            <h4 class="date">2017-01-27</h4>
          </div>

    
    
<div class="contents">
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#overview" class="anchor"> </a></body></html>Overview</h1>
<p>We will learn how to use the package <a href="http://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5">MAST</a> to analyze single cell RNAseq experiments.</p>
<p>Starting from a matrix of counts of transcripts (cells by transcripts), we will discuss the preliminary steps of quality control, filtering, and exploratory data analysis. Once we are satisfied that we have high-quality expression, we will consider tests for differential expression and ways to visualize results. It is often helpful to synthesize from gene-level into module-level statements. Therefore, we will learn how to use MAST to test for gene set enrichment.</p>
<p>We will apply these methods to a data set of Mucosal Associated Invariant T cells (MAITs), measured on the Fluidigm C1. Half of the cells have been cytokine stimulated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">library</span>(BiocInstaller)
<span class="kw">useDevel</span>()
<span class="kw">biocLite</span>(<span class="st">'SummarizedExperiment'</span>)
<span class="kw">biocValid</span>()
<span class="kw">install_github</span>(<span class="st">'RGLab/MAST@summarizedExpt'</span>)</code></pre></div>
<p>If you are running this at BioC, you should not need to run this chunk. Otherwise you will need to install the summarizedExpt branch and the devel (3.4) branch of Bioconductor.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">suppressPackageStartupMessages</span>({
    <span class="kw">library</span>(ggplot2)
    <span class="kw">library</span>(GGally)
    <span class="kw">library</span>(GSEABase)
    <span class="kw">library</span>(limma)
    <span class="kw">library</span>(reshape2)
    <span class="kw">library</span>(data.table)
    <span class="kw">library</span>(knitr)
    <span class="kw">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)
    <span class="kw">library</span>(stringr)
    <span class="kw">library</span>(NMF)
    <span class="kw">library</span>(rsvd)
    <span class="kw">library</span>(RColorBrewer)
    <span class="kw">library</span>(MAST)
})
<span class="co">#options(mc.cores = detectCores() - 1) #if you have multiple cores to spin</span>
<span class="kw">options</span>(<span class="dt">mc.cores =</span> <span class="dv">1</span>)
knitr::opts_chunk$<span class="kw">set</span>(<span class="dt">message =</span> <span class="ot">FALSE</span>,<span class="dt">error =</span> <span class="ot">FALSE</span>,<span class="dt">warning =</span> <span class="ot">FALSE</span>,<span class="dt">cache =</span> <span class="ot">TRUE</span>,<span class="dt">fig.width=</span><span class="dv">8</span>,<span class="dt">fig.height=</span><span class="dv">6</span>, <span class="dt">auto.dep=</span><span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="loading-and-transforming-data" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#loading-and-transforming-data" class="anchor"> </a></body></html>Loading and transforming data</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">freq_expressed &lt;-<span class="st"> </span><span class="fl">0.2</span>
FCTHRESHOLD &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="fl">1.5</span>)</code></pre></div>
<p>First, let’s set some constants that we can use throughout this analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(maits, <span class="dt">package=</span><span class="st">'MAST'</span>)
<span class="kw">dim</span>(maits$expressionmat)
<span class="kw">head</span>(maits$cdat)
<span class="kw">head</span>(maits$fdat)</code></pre></div>
<p>Next, let’s load the data, which consists of a matrix of log2 + 1 transcripts per million (TPM), as output by RSEM. Internally, we use the packages <a href="https://github.com/RGLab/RNASeqPipelineR">RNASeqPipelineR</a> and <a href="https://github.com/RGLab/preProcessData">preprocessData</a> to facilitate aligning and quantitating the raw reads. This is an experiment on Mucosal Associated Invariant T-cells from a single healthy donor. A number of cells were subjected to reverse transcription and library prep immediately, while others were first stimulated for 24 hours with a cytokine cocktail.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scaRaw &lt;-<span class="st"> </span><span class="kw"><a href="../reference/FromMatrix.html">FromMatrix</a></span>(<span class="kw">t</span>(maits$expressionmat), maits$cdat, maits$fdat)</code></pre></div>
<p><code>FromMatrix</code> constructs an object from a matrix (genes <span class="math inline">\(\times\)</span> cells) of expression, a <code>data.frame</code> of cell-level covariance and a <code>data.frame</code> of feature-level covariates. In this case, there are 96 single cells measured over 16302 genes. We derive from the <code>SummarizedExperiment</code> class, which makes it easy for feature (row) and cell (column) metadata to come along for the ride. There is also a constructor, <code>FromFlatDF</code> for flattened data where measurements, and cell and feature covariates are intermingled.</p>
<div id="exploratory-data-analysis" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#exploratory-data-analysis" class="anchor"> </a></body></html>Exploratory Data Analysis</h2>
<p>Let’s explore these data with a heatmap and some PCA.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">aheatmap</span>(<span class="kw">assay</span>(scaRaw[<span class="dv">1</span>:<span class="dv">1000</span>,]), <span class="dt">labRow=</span><span class="st">''</span>, <span class="dt">annCol=</span><span class="kw">as.data.frame</span>(<span class="kw">colData</span>(scaRaw)[,<span class="kw">c</span>(<span class="st">'condition'</span>, <span class="st">'ourfilter'</span>)]), <span class="dt">distfun=</span><span class="st">'spearman'</span>)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/heatmap-1.png" width="768"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">123</span>)
plotPCA &lt;-<span class="st"> </span>function(sca_obj){
    projection &lt;-<span class="st"> </span><span class="kw">rpca</span>(<span class="kw">t</span>(<span class="kw">assay</span>(sca_obj)), <span class="dt">retx=</span><span class="ot">TRUE</span>, <span class="dt">k=</span><span class="dv">4</span>)$x
    pca &lt;-<span class="st"> </span><span class="kw">data.table</span>(projection,  <span class="kw">as.data.frame</span>(<span class="kw">colData</span>(sca_obj)))
    <span class="kw">print</span>(<span class="kw">ggpairs</span>(pca, <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">'PC1'</span>, <span class="st">'PC2'</span>, <span class="st">'PC3'</span>, <span class="st">'libSize'</span>, <span class="st">'PercentToHuman'</span>, <span class="st">'nGeneOn'</span>, <span class="st">'exonRate'</span>),
            <span class="dt">mapping=</span><span class="kw">aes</span>(<span class="dt">color=</span>condition), <span class="dt">upper=</span><span class="kw">list</span>(<span class="dt">continuous=</span><span class="st">'blank'</span>)))
    <span class="kw">invisible</span>(pca)
}

<span class="kw">plotPCA</span>(scaRaw)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/pca-1.png" width="768"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filterCrit &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">colData</span>(scaRaw), pastFastqc==<span class="st">"PASS"</span>&amp;<span class="st"> </span>exonRate &gt;<span class="fl">0.3</span> &amp;<span class="st"> </span>PercentToHuman&gt;<span class="fl">0.6</span> &amp;<span class="st"> </span>nGeneOn&gt;<span class="st"> </span><span class="dv">4000</span>)</code></pre></div>
<p>From the PCA and heatmap we can tell that we’ve got some data quality issues with low-diversity libraries, and libraries where we capture a lot of non mRNA. We’ll set a filtering threshold using the <a href="https://en.wikipedia.org/wiki/I_know_it_when_I_see_it">Potter Stewart</a> method, though Shalek et al and others have come up with more principled ways, such as training SVM on the expression data and quality metrics to predict failed libraries. Though this still presupposes a set of labeled “failures” and “successes”.</p>
</div>
<div id="filtering" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#filtering" class="anchor"> </a></body></html>Filtering</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sca &lt;-<span class="st"> </span><span class="kw">subset</span>(scaRaw,filterCrit)
eid &lt;-<span class="st"> </span><span class="kw">select</span>(TxDb.Hsapiens.UCSC.hg19.knownGene,<span class="dt">keys =</span> <span class="kw">mcols</span>(sca)$entrez,<span class="dt">keytype =</span><span class="st">"GENEID"</span>,<span class="dt">columns =</span> <span class="kw">c</span>(<span class="st">"GENEID"</span>,<span class="st">"TXNAME"</span>))
ueid &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">na.omit</span>(eid)$GENEID)
sca &lt;-<span class="st"> </span>sca[<span class="kw">mcols</span>(sca)$entrez %in%<span class="st"> </span>ueid,]
## Remove invariant genes
sca &lt;-<span class="st"> </span>sca[<span class="kw">sample</span>(<span class="kw">which</span>(<span class="kw"><a href="../reference/freq.html">freq</a></span>(sca)&gt;<span class="dv">0</span>), <span class="dv">6000</span>),]</code></pre></div>
<p>We’ll now consider only cells that pass the filter. <code>subset(scaRaw, filterCrit)</code> is just syntactic sugar for subsetting by columns, but who doesn’t like a little sugar? We’ll also limit ourselves to transcripts that have an EntrezGene id to facilitate interpretation. Lastly, we will take a subsample of about 50% of the total genes to keep this vignette from taking too long or too much memory.</p>
<div id="recalculating-the-cellular-detection-rate-ngeneson" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#recalculating-the-cellular-detection-rate-ngeneson" class="anchor"> </a></body></html>Recalculating the cellular detection rate (ngeneson)</h3>
<p>We and others have found that the number of genes detected in a sample, which we deemed the cellular detection rate is often the first principal component. After we removed genes low expression, perhaps we want to recalculate it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cdr2 &lt;-<span class="kw">colSums</span>(<span class="kw">assay</span>(sca)&gt;<span class="dv">0</span>)
<span class="kw">qplot</span>(<span class="dt">x=</span>cdr2, <span class="dt">y=</span><span class="kw">colData</span>(sca)$nGeneOn) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">'New CDR'</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="st">'Old CDR'</span>)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/unnamed-chunk-2-1.png" width="384"></p>
<p>We can assign this to a new column in the <code>colData</code> in the obvious fashion:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colData</span>(sca)$cngeneson &lt;-<span class="st"> </span><span class="kw">scale</span>(cdr2)</code></pre></div>
</div>
<div id="pca-on-filtered-cells" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#pca-on-filtered-cells" class="anchor"> </a></body></html>PCA on filtered cells</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(sca)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/pcaFilter-1.png" width="768"> PC1 now primarily captures stimulation effects. We still observe that PC2 correlates with the exonRate and ngeneson, but the data are more smoothly distributed.</p>
</div>
<div id="exercises-on-loading-and-transforming-data" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#exercises-on-loading-and-transforming-data" class="anchor"> </a></body></html>Exercises on loading and transforming data</h3>
<ol style="list-style-type: decimal">
<li>These samples were reverse-transcribed and amplified on two Fluidigm C1 chips (and stimulation is completely confounded with chip…) The chip location is contained in the column <code>wellKey</code> in <code>colData(sca)</code> as a three-character string like <code>CXX</code>. How could you extract the chip location and add it as column to the cellular data? Hint: <code>str_split_fixed</code> or <code>str_extract</code> may be your friends here.</li>
</ol>
</div>
</div>
</div>
<div id="adaptive-thresholding" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#adaptive-thresholding" class="anchor"> </a></body></html>Adaptive thresholding</h1>
<p>Single cell gene expression data are known to be zero-inflated and bimodal, which is feature we observe here as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scaSample &lt;-<span class="st"> </span>sca[<span class="kw">sample</span>(<span class="kw">which</span>(<span class="kw"><a href="../reference/freq.html">freq</a></span>(sca)&gt;.<span class="dv">1</span>), <span class="dv">20</span>),]
flat &lt;-<span class="st"> </span><span class="kw">as</span>(scaSample, <span class="st">'data.table'</span>)
<span class="kw">ggplot</span>(flat, <span class="kw">aes</span>(<span class="dt">x=</span>value))+<span class="kw">geom_density</span>() +<span class="kw">facet_wrap</span>(~symbolid, <span class="dt">scale=</span><span class="st">'free_y'</span>)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/distribution-1.png" width="576"></p>
<p>Here we’ve taken subsample of size 20, then flattened it into a <code>data.table</code> to make it easy to plot with <code>ggplot2</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">thres &lt;-<span class="st"> </span><span class="kw"><a href="../reference/thresholdSCRNACountMatrix.html">thresholdSCRNACountMatrix</a></span>(<span class="kw">assay</span>(sca), <span class="dt">nbins =</span> <span class="dv">20</span>, <span class="dt">min_per_bin =</span> <span class="dv">30</span>)
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">4</span>))
<span class="kw">plot</span>(thres)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/threshold-1.png" width="576"></p>
<p>We suspect that the left-most mode corresponds to non-specific hybridization of mRNA or genomic DNA. Here we apply an adaptive scheme to threshold values below a cut-off that depends on the intensity of the signal cluster from the gene (determined from the median expression value). When we plot the threshold vs genes binned by median expression value, we can see this evolution of thresholding value.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assays</span>(sca) &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">thresh=</span>thres$counts_threshold, <span class="dt">tpm=</span><span class="kw">assay</span>(sca))
expressed_genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/freq.html">freq</a></span>(sca) &gt;<span class="st"> </span>freq_expressed
sca &lt;-<span class="st"> </span>sca[expressed_genes,]</code></pre></div>
<p>We’ll limit ourselves to genes that are found in at least 0.2 of the sample (since we won’t have any power to conclude much about less frequent transcripts). We will also downsample to 1000 genes to keep this vignette from running too slowly.</p>
<div id="exercises-on-thresholding" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#exercises-on-thresholding" class="anchor"> </a></body></html>Exercises on thresholding</h3>
<ol start="2" style="list-style-type: decimal">
<li>Many modeling procedures assume (or perform best) when the data are Normally distributed. How could we evaluate how thresholding affects the Normality of the data? How about the Normality of the non-zero data?</li>
</ol>
</div>
</div>
<div id="differential-expression-using-a-hurdle-model" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#differential-expression-using-a-hurdle-model" class="anchor"> </a></body></html>Differential Expression using a Hurdle model</h1>
<p>We’ll fit a hurdle model, modeling the condition and (centered) <code>ngeneson</code> factor, thus adjusting for the cellular detection rate.</p>
<p>In order to have more interpretable coefficients, we’ll set the reference level of the factor to be the “unstimulated” cells.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cond&lt;-<span class="kw">factor</span>(<span class="kw">colData</span>(sca)$condition)
cond&lt;-<span class="kw">relevel</span>(cond,<span class="st">"Unstim"</span>)
<span class="kw">colData</span>(sca)$condition&lt;-cond
zlmCond &lt;-<span class="st"> </span><span class="kw"><a href="../reference/zlm.html">zlm</a></span>(~condition +<span class="st"> </span>cngeneson, sca)
<span class="co"># The following are equivalent</span>
## lrt &lt;- lrTest(zlm, "condition")
## lrt &lt;- lrTest(zlm, CoefficientHypothesis('conditionStim'))

<span class="co"># This would test if 2*cngeneson=conditionStim</span>
<span class="co">#  This is sheer nonsense biologically and statistically, but gives an example of the flexibility.</span>
## lrt &lt;- lrTest(zlm, Hypothesis('2*cngeneson-conditionStim'))</code></pre></div>
<p>We could run a likelihood ratio test here, testing for differences when we drop the <code>condition</code> factor. Note that any arbitrary contrast matrix can be tested here, and specified either using a matrix or syntactically. See <code>Hypothesis</code> for details.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#only test the condition coefficient.</span>
summaryCond &lt;-<span class="st"> </span><span class="kw">summary</span>(zlmCond, <span class="dt">doLRT=</span><span class="st">'conditionStim'</span>) 
<span class="co">#print the top 4 genes by contrast using the logFC</span>
<span class="kw">print</span>(summaryCond, <span class="dt">n=</span><span class="dv">4</span>)</code></pre></div>
<pre><code>## Fitted zlm with top 4 genes per contrast:
## ( log fold change Z-score )
##  primerid conditionStim cngeneson
##  1327        0.9           3.8*  
##  255252     -5.1           4.1*  
##  285962    -10.0*          0.4   
##  3727       -7.8*          0.5   
##  392490     -1.1          -4.3*  
##  7295        7.7*          0.4   
##  84314      -7.0*          1.7   
##  9262       -2.1           3.8*</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## by discrete Z-score
<span class="kw">print</span>(summaryCond, <span class="dt">n=</span><span class="dv">4</span>, <span class="dt">by=</span><span class="st">'D'</span>)</code></pre></div>
<pre><code>## Fitted zlm with top 4 genes per contrast:
## ( Wald Z-scores on discrete )
##  primerid  conditionStim cngeneson
##  100132356   -4.0*          2.2   
##  100506548   -2.6           4.2*  
##  134147      -4.1*          3.1   
##  163702      -3.4           4.5*  
##  26024        4.0*         -0.7   
##  285962      -4.3*          1.4   
##  56996       -3.5           4.5*  
##  6360        -3.1           4.4*</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## by continuous Z-score
<span class="kw">print</span>(summaryCond, <span class="dt">n=</span><span class="dv">4</span>, <span class="dt">by=</span><span class="st">'C'</span>)</code></pre></div>
<pre><code>## Fitted zlm with top 4 genes per contrast:
## ( Wald Z-scores tests on continuous )
##  primerid conditionStim cngeneson
##  152926      6.0*         -2.9   
##  3727       -5.9*          0.4   
##  514         3.1          -4.4*  
##  54964       1.9          -4.0*  
##  6232       -6.1*         -0.7   
##  6482        3.8          -5.7*  
##  7295        5.9*         -1.2   
##  9459        3.2          -4.0*</code></pre>
<p>But often of more general use is this delicious syntactic sugar to make a giant <code>data.table</code> containing coefficients, standard errors, etc, for the various model components. Many Bothan spies died so that we could pretty-print this summary of the top differentially expressed genes.</p>
<p>Strip off the <code>$datatable</code> component to stop this pretty-printing (or call <code>print.default</code>.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">summaryDt &lt;-<span class="st"> </span>summaryCond$datatable
fcHurdle &lt;-<span class="st"> </span><span class="kw">merge</span>(summaryDt[contrast==<span class="st">'conditionStim'</span> &amp;<span class="st"> </span>component==<span class="st">'H'</span>,.(primerid, <span class="st">`</span><span class="dt">Pr(&gt;Chisq)</span><span class="st">`</span>)], <span class="co">#hurdle P values</span>
                      summaryDt[contrast==<span class="st">'conditionStim'</span> &amp;<span class="st"> </span>component==<span class="st">'logFC'</span>, .(primerid, coef, ci.hi, ci.lo)], <span class="dt">by=</span><span class="st">'primerid'</span>) <span class="co">#logFC coefficients</span>

fcHurdle[,fdr:<span class="er">=</span><span class="kw">p.adjust</span>(<span class="st">`</span><span class="dt">Pr(&gt;Chisq)</span><span class="st">`</span>, <span class="st">'fdr'</span>)]</code></pre></div>
<pre><code>##        primerid Pr(&gt;Chisq)        coef       ci.hi       ci.lo         fdr
##    1: 100093630 0.92842013 -0.52873971  2.62207632 -3.67955574 0.953040341
##    2:     10010 0.22229015  3.14003456  6.32251110 -0.04244198 0.402246495
##    3: 100128025 0.00021654 -0.95005072 -0.37952708 -1.52057436 0.006929279
##    4: 100128398 0.00311580 -1.35056067 -0.59032243 -2.11079891 0.031972437
##    5: 100128675 0.04251842 -1.06436962  0.03623994 -2.16497918 0.141531511
##   ---                                                                     
## 2396:      9976 0.40295544  0.08534798  2.49599376 -2.32529780 0.575993489
## 2397:       998 0.22156699  2.31249593  5.06949104 -0.44449918 0.401940638
## 2398:      9984 0.35132082 -1.45791254  0.83185969 -3.74768476 0.531403817
## 2399:      9988 0.40731378  0.71706845  2.53070606 -1.09656917 0.579806086
## 2400:      9991 0.77580882 -0.37921891  1.84377992 -2.60221774 0.842126261</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fcHurdleSig &lt;-<span class="st"> </span><span class="kw">merge</span>(fcHurdle[fdr&lt;.<span class="dv">05</span> &amp;<span class="st"> </span><span class="kw">abs</span>(coef)&gt;FCTHRESHOLD], <span class="kw">as.data.table</span>(<span class="kw">mcols</span>(sca)), <span class="dt">by=</span><span class="st">'primerid'</span>)
<span class="kw">setorder</span>(fcHurdleSig, fdr)</code></pre></div>
<p>We see that there are 279 genes significant at a FDR of 5% and with a log-fold change larger than 0.6.</p>
<div id="visualization-of-50-most-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#visualization-of-50-most-differentially-expressed-genes" class="anchor"> </a></body></html>Visualization of 50 most differentially expressed genes</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">entrez_to_plot &lt;-<span class="st"> </span>fcHurdleSig[<span class="dv">1</span>:<span class="dv">50</span>,primerid]
symbols_to_plot &lt;-<span class="st"> </span>fcHurdleSig[<span class="dv">1</span>:<span class="dv">50</span>,symbolid]
flat_dat &lt;-<span class="st"> </span><span class="kw">as</span>(sca[entrez_to_plot,], <span class="st">'data.table'</span>)
ggbase &lt;-<span class="st"> </span><span class="kw">ggplot</span>(flat_dat, <span class="kw">aes</span>(<span class="dt">x=</span>condition, <span class="dt">y=</span>thresh,<span class="dt">color=</span>condition)) +<span class="st"> </span><span class="kw">geom_jitter</span>()+<span class="kw">facet_wrap</span>(~symbolid, <span class="dt">scale=</span><span class="st">'free_y'</span>)+<span class="kw">ggtitle</span>(<span class="st">"DE Genes in Activated MAIT Cells"</span>)
ggbase+<span class="kw">geom_violin</span>() </code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/unnamed-chunk-5-1.png" width="768"> Here is a standard violin plot showing how expression differs in each gene between stimulated and unstimulated cells. What could be improved about this visualization to make it more accurately reflect the model that’s being fit?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">flat_dat[,lmPred:<span class="er">=</span><span class="kw">lm</span>(thresh~cngeneson +<span class="st"> </span>condition)$fitted, key=symbolid]</code></pre></div>
<pre><code>##       primerid entrez symbolid               wellKey condition nGeneOn
##    1:    83451  83451   ABHD11   1-MAIT-Stim-C05_S91      Stim    8805
##    2:    83451  83451   ABHD11   1-MAIT-Stim-C07_S70      Stim    9184
##    3:    83451  83451   ABHD11   1-MAIT-Stim-C08_S68      Stim    8895
##        libSize PercentToHuman MedianCVCoverage PCRDuplicate exonRate
##    1: 10055037          0.896         0.938330     0.756202    0.651
##    2: 13822410          0.896         0.687973     0.788929    0.764
##    3: 12122964          0.891         0.947196     0.799650    0.706
##       pastFastqc ncells  ngeneson  cngeneson     TRAV1     TRBV6 TRBV4
##    1:       PASS      1 0.2764538  0.9510350 10.390061  0.000000     0
##    2:       PASS      1 0.2882285  1.3203917  7.639777  0.000000     0
##    3:       PASS      1 0.2789186  0.9870698  8.657032  0.000000     0
##       TRBV20 alpha  beta           ac           bc ourfilter   thresh
##    1:      0 TRAV1    NA   TRAV1.Stim           NA      TRUE 3.571677
##    2:      0 TRAV1    NA   TRAV1.Stim           NA      TRUE 5.163096
##    3:      0 other    NA   other.Stim           NA      TRUE 2.784504
##             tpm   lmPred
##    1: 3.5716768 2.119838
##    2: 5.1630961 2.567200
##    3: 2.7845040 2.163483
##  [ reached getOption("max.print") -- omitted 8 rows ]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ggbase +<span class="kw">aes</span>(<span class="dt">x=</span>cngeneson) +<span class="st"> </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">y=</span>lmPred), <span class="dt">lty=</span><span class="dv">1</span>) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">'Standardized Cellular Detection Rate'</span>)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/unnamed-chunk-6-1.png" width="768"> Really we are doing an analysis of covariance (ANCOVA), where we ask if at each value of <code>cngeneson</code> if the stimulated cells differ from the unstimulated, in multiple module components. That is a little bit complicated to visualize, so we’ll settle for just showing what the ANCOVA looks like on the zero-inflated data. Adding these fitted values is easy with <code>data.table</code> on the melted data.</p>
</div>
<div id="visualizing-both-components" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#visualizing-both-components" class="anchor"> </a></body></html>Visualizing both components</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## This is all rather kludgy at the moment
MM &lt;-<span class="st"> </span><span class="kw"><a href="../reference/model.matrix.html">model.matrix</a></span>(~condition,<span class="kw">unique</span>(<span class="kw">colData</span>(sca)[,<span class="kw">c</span>(<span class="st">"condition"</span>),<span class="dt">drop=</span><span class="ot">FALSE</span>]))
<span class="kw">rownames</span>(MM) &lt;-<span class="st"> </span><span class="kw">str_extract</span>(<span class="kw">rownames</span>(MM), <span class="st">'Stim|Unstim'</span>)
predicted &lt;-<span class="st"> </span><span class="kw">predict</span>(zlmCond,<span class="dt">modelmatrix=</span>MM)

## Avert your eyes...
predicted[, primerid:<span class="er">=</span><span class="kw">as.character</span>(primerid)]</code></pre></div>
<pre><code>##        primerid sample      muC       seC         muD       seD
##    1: 100093630   Stim 7.450230 0.8626787 -0.14368837 0.4531873
##    2: 100093630 Unstim 7.651772 0.7997469  0.08410861 0.4422192
##    3:     10010   Stim 7.616231 0.9402944  0.82895280 0.5740620
##    4:     10010 Unstim 7.717271 1.1383388 -0.94379137 0.5465047
##    5: 100128025   Stim 1.469193 0.6511569 -3.27455722 0.7318365
##   ---                                                          
## 4796:      9984 Unstim 6.740629 1.2783640 -0.62921710 0.4763362
## 4797:      9988   Stim 4.024973 1.0406955 -0.42189487 0.4984322
## 4798:      9988 Unstim 5.003977 1.2216902 -1.54868836 0.5559121
## 4799:      9991   Stim 5.522831 1.0471248 -1.00376664 0.5025482
## 4800:      9991 Unstim 6.734100 1.0367372 -0.96302440 0.4916566</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">predicted_sig &lt;-<span class="st"> </span><span class="kw">merge</span>(<span class="kw">mcols</span>(sca), predicted[primerid%in%entrez_to_plot], <span class="dt">by=</span><span class="st">'primerid'</span>)
predicted_sig &lt;-<span class="st"> </span><span class="kw">as.data.table</span>(predicted_sig)

## plot with inverse logit transformed x-axis
<span class="kw">ggplot</span>(predicted_sig)+<span class="kw">aes</span>(<span class="dt">x=</span><span class="kw"><a href="../reference/invlogit.html">invlogit</a></span>(muD),<span class="dt">y=</span>muC,<span class="dt">xse=</span>seD,<span class="dt">yse=</span>seC,<span class="dt">col=</span>sample)+
<span class="st">    </span><span class="kw">facet_wrap</span>(~symbolid,<span class="dt">scales=</span><span class="st">"free_y"</span>)+<span class="kw">theme_linedraw</span>()+
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="fl">0.5</span>)+<span class="kw">scale_x_continuous</span>(<span class="st">"Proportion expression"</span>)+
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="st">"Estimated Mean"</span>)+
<span class="st">    </span><span class="kw"><a href="../reference/stat_ell.html">stat_ell</a></span>(<span class="kw">aes</span>(<span class="dt">x=</span>muD,<span class="dt">y=</span>muC),<span class="dt">level=</span><span class="fl">0.95</span>, <span class="dt">invert=</span><span class="st">'x'</span>)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/unnamed-chunk-7-1.png" width="768"> Lastly, we can visualize how each component of the hurdle model is combined to yield a differential test by plotting the estimated coefficients (and standard errors) as a scatter plot. These confidence ellipses are drawn based on the <span class="math inline">\(\chi^2_2\)</span> distribution on the simultaneous test that either coefficient differs from zero. What do you notice about the relationship between the confidence interval on the proportion (x) and mean (y)?</p>
</div>
<div id="heatmap-of-maits-based-on-most-differentially-expressed-genes" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#heatmap-of-maits-based-on-most-differentially-expressed-genes" class="anchor"> </a></body></html>Heatmap of MAITs based on most differentially expressed genes</h2>
<p>Here’s what the 50 most significantly differentially expressed(DE) genes look like.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat_to_plot &lt;-<span class="st"> </span><span class="kw">assay</span>(sca[entrez_to_plot,])
<span class="kw">rownames</span>(mat_to_plot) &lt;-<span class="st"> </span>symbols_to_plot
<span class="kw">aheatmap</span>(mat_to_plot,<span class="dt">annCol=</span><span class="kw">colData</span>(sca)[,<span class="st">"condition"</span>],<span class="dt">main=</span><span class="st">"DE genes"</span>,<span class="dt">col=</span><span class="kw">rev</span>(<span class="kw">colorRampPalette</span>(<span class="dt">colors =</span> <span class="kw">brewer.pal</span>(<span class="dt">name=</span><span class="st">"PiYG"</span>,<span class="dt">n=</span><span class="dv">10</span>))(<span class="dv">20</span>)))</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/unnamed-chunk-8-1.png" width="768"> Some of the activated MAITs have transcriptional profiles more similar to unactivated MAITS.</p>
<div id="exercises-on-hurdle-model-differential-expression" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#exercises-on-hurdle-model-differential-expression" class="anchor"> </a></body></html>Exercises on hurdle model differential expression</h3>
<ol start="3" style="list-style-type: decimal">
<li>In some of the libraries it was possible to recover portions of the T-cell receptor alpha and beta chains from the reads. Behold, we live a golden age:</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">colData</span>(sca)$beta, <span class="dt">exclude=</span><span class="ot">NULL</span>)</code></pre></div>
<pre><code>## 
##  other TRBV20  TRBV4  TRBV6   &lt;NA&gt; 
##      9      5      4     18     37</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Note that we currently throw an uninformative error if a covariate is `NA`</span>
scaHasBeta &lt;-<span class="st"> </span><span class="kw">subset</span>(sca, !<span class="kw">is.na</span>(beta))</code></pre></div>
<p>How could you test for differential expression between TCR <code>beta</code> subtypes among <code>scaHasBeta</code>, the cells in which the TCR was successfully inferred? What is worrying about this procedure?</p>
</div>
</div>
<div id="residuals" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#residuals" class="anchor"> </a></body></html>Residuals</h2>
<p>In the MAST paper, we assessed this further by examining a <em>residual</em> from the Hurdle linear model. This is not an entirely well-defined concept, as there are really two residuals in the Hurdle model, and there are multiple forms of residual that one can consider in logistic regression. Nonetheless, it can be a useful diagnostic. You can get a form of <a href="https://en.wikipedia.org/wiki/Deviance_(statistics)">deviance residual</a> by supplying a “hook” function to be called after each regression.</p>
<p>For example, in the subset of differentially expressed genes, we can refit the model and pull out the residual.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scaDE &lt;-<span class="st"> </span>sca[entrez_to_plot,]
zlmResidDE &lt;-<span class="st"> </span><span class="kw"><a href="../reference/zlm.html">zlm</a></span>(~condition +<span class="st"> </span>cngeneson, scaDE, <span class="dt">hook=</span>deviance_residuals_hook)
residDE &lt;-<span class="st"> </span>zlmResidDE@hookOut
residDEMatrix &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, residDE)</code></pre></div>
<p>The <code>residDE</code> was a list of residuals (one per gene in <code>scaDE</code>). We concatenated them into a matrix.</p>
<p>Now we can add <code>residDEMatrix</code> as another component of the “assay”:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assays</span>(scaDE) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">assays</span>(scaDE), <span class="kw">list</span>(<span class="dt">resid=</span>residDEMatrix))
scaResidFlat &lt;-<span class="st"> </span><span class="kw">as</span>(scaDE, <span class="st">'data.table'</span>)
scaResidFlat[<span class="dv">1</span>:<span class="dv">4</span>,]</code></pre></div>
<pre><code>##    primerid entrez symbolid             wellKey condition nGeneOn  libSize
## 1:     3727   3727     JUND 1-MAIT-Stim-C05_S91      Stim    8805 10055037
## 2:     7295   7295      TXN 1-MAIT-Stim-C05_S91      Stim    8805 10055037
## 3:   285962 285962 WEE2-AS1 1-MAIT-Stim-C05_S91      Stim    8805 10055037
##    PercentToHuman MedianCVCoverage PCRDuplicate exonRate pastFastqc ncells
## 1:          0.896          0.93833     0.756202    0.651       PASS      1
## 2:          0.896          0.93833     0.756202    0.651       PASS      1
## 3:          0.896          0.93833     0.756202    0.651       PASS      1
##     ngeneson cngeneson    TRAV1 TRBV6 TRBV4 TRBV20 alpha beta         ac
## 1: 0.2764538  0.951035 10.39006     0     0      0 TRAV1   NA TRAV1.Stim
## 2: 0.2764538  0.951035 10.39006     0     0      0 TRAV1   NA TRAV1.Stim
## 3: 0.2764538  0.951035 10.39006     0     0      0 TRAV1   NA TRAV1.Stim
##    bc ourfilter   thresh        tpm      resid
## 1: NA      TRUE  4.76288  4.7628803  0.7056185
## 2: NA      TRUE 10.57847 10.5784670  0.1036310
## 3: NA      TRUE  0.00000  1.5109619 -0.6147637
##  [ reached getOption("max.print") -- omitted 1 row ]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(scaResidFlat, <span class="kw">aes</span>(<span class="dt">x=</span>ngeneson, <span class="dt">y=</span>resid))+<span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">col=</span>condition))+<span class="kw">geom_smooth</span>()+<span class="kw">facet_wrap</span>(~symbolid)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/addResiduals-1.png" width="768"> This allows us, for example, to evaluate model diagnostics such as this “residuals vs variable” plot to check the linearity assumption of “ngeneson”. One can also conduct PCA or examine the correlation matrix to look for co-expressing modules of genes.</p>
</div>
</div>
<div id="gene-set-enrichment-analysis" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#gene-set-enrichment-analysis" class="anchor"> </a></body></html>Gene Set Enrichment Analysis</h1>
<p>We use a competitive gene set enrichment test, in which a contrast (hurdle model coefficient) from various gene sets of interest is compared to the background, accounting for the intergene correlation of the module coefficient.</p>
<p>To estimate the intergene correlation of the contrast, we use bootstrapping. Cells (columns in <code>sca</code>) are sampled with replacement a number of times and we refit the model. At some point, we may implement a GEE-like method to estimate the intergene correlation through asymptotic calculations. But in the meantime, the bootstrapping can be slow, although is pleasingly parallelizable. To avoid setting the server farm alight, we will work with only a subset of the genes, and only calculate 5 bootstrap replicates. This would be completely inadequate if you wanted to report these results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># bootstrap, resampling cells</span>
<span class="co"># R should be set to &gt;50 if you were doing this for real.</span>
boots &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bootVcov1.html">bootVcov1</a></span>(zlmCond, <span class="dt">R=</span><span class="dv">4</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">module &lt;-<span class="st"> "BTM"</span>
min_gene_in_module &lt;-<span class="st"> </span><span class="dv">5</span>
packageExt &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package=</span><span class="st">'MAST'</span>)
module_file &lt;-<span class="st"> </span><span class="kw">list.files</span>(packageExt, <span class="dt">pattern =</span> module, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)
gene_set &lt;-<span class="st"> </span><span class="kw">getGmt</span>(module_file)
gene_ids &lt;-<span class="st"> </span><span class="kw">geneIds</span>(gene_set)
gene_ids &lt;-<span class="st"> </span>gene_ids[!<span class="kw">names</span>(gene_ids)%like%<span class="st">"TBA"</span>&amp;!<span class="kw">names</span>(gene_ids)%like%<span class="st">"B cell"</span>]
sets_indices &lt;-<span class="st"> </span>limma::<span class="kw">ids2indices</span>(gene_ids, <span class="kw">mcols</span>(sca)$symbolid)
<span class="co"># Only keep modules with at least min_gene_in_module</span>
sets_indices &lt;-<span class="st"> </span>sets_indices[<span class="kw">sapply</span>(sets_indices, length) &gt;=<span class="st"> </span>min_gene_in_module]</code></pre></div>
<p>We will use the Blood Transcriptional Modules of <a href="www.nature.com/ni/journal/v15/n2/full/ni.2789.html">Li, et al</a> and select named modules not relating to B-cell functionality.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gsea &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gseaAfterBoot.html">gseaAfterBoot</a></span>(zlmCond, boots, sets_indices, <span class="kw"><a href="../reference/Hypothesis.html">CoefficientHypothesis</a></span>(<span class="st">"conditionStim"</span>)) 
z_stat_comb &lt;-<span class="st"> </span><span class="kw">summary</span>(gsea, <span class="dt">testType=</span><span class="st">'normal'</span>)</code></pre></div>
<p>The <code>summary</code> method returns a <code>data.table</code> with columns giving discrete and continuous Z-scores (<code>disc_Z</code> and <code>cont_Z</code>) and P-values testing if the average coefficient in the gene set differs from the average coefficient outside the set. A combined P-value (using Stouffer’s method) is given in column <code>combined_P</code>. The effect sizes (difference in average regression coefficients) is given in <code>effect_disc</code> and <code>effect_cont</code>. For the discrete component this gives, for example, the difference in the average odds of expression in the set vs outside the set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sigModules &lt;-<span class="st"> </span>z_stat_comb[combined_adj&lt;.<span class="dv">01</span>]
gseaTable &lt;-<span class="st"> </span><span class="kw">melt</span>(sigModules[,.(set, disc_Z, cont_Z, combined_Z)], <span class="dt">id.vars=</span><span class="st">'set'</span>)
<span class="kw">ggplot</span>(gseaTable, <span class="kw">aes</span>(<span class="dt">y=</span>set, <span class="dt">x=</span>variable, <span class="dt">fill=</span>value))+<span class="kw">geom_raster</span>() +<span class="st"> </span><span class="kw">scale_fill_distiller</span>(<span class="dt">palette=</span><span class="st">"PiYG"</span>)</code></pre></div>
<p><img src="MAITAnalysis_files/figure-html/gseaView-1.png" width="768"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li>
<a href="#loading-and-transforming-data">Loading and transforming data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#exploratory-data-analysis">Exploratory Data Analysis</a></li>
      <li><a href="#filtering">Filtering</a></li>
      </ul>
</li>
      <li><a href="#adaptive-thresholding">Adaptive thresholding</a></li>
      <li>
<a href="#differential-expression-using-a-hurdle-model">Differential Expression using a Hurdle model</a><ul class="nav nav-pills nav-stacked">
<li><a href="#heatmap-of-maits-based-on-most-differentially-expressed-genes">Heatmap of MAITs based on most differentially expressed genes</a></li>
      <li><a href="#residuals">Residuals</a></li>
      </ul>
</li>
      <li><a href="#gene-set-enrichment-analysis">Gene Set Enrichment Analysis</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Andrew McDavid, Greg Finak, Masanao Yajima.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
