<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ MAST</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MAST</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/RGLab/MAST/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9">
    <div class="page-header">
      <h1>Function reference</h1>
      <h4>version&nbsp;1.0.6</h4>
    </div>

    <div class="contents">
      <h2 id="section-loading-transforming-and-summarizing" class="hasAnchor"><a href="#section-loading-transforming-and-summarizing" class="anchor"></a>Loading, transforming and summarizing</h2>
      <p class="section-desc"><p>Functions to load data, add covariates or annotations, or provide high-level summaries.</p></p>

      
      <p><a href="convertMASTClassicToSingleCellAssay.html">Convert a MASTClassic SingleCellAssay</a></p>
      <ul>
        <li><code><a href="convertMASTClassicToSingleCellAssay.html">convertMASTClassicToSingleCellAssay</a></code></li>
      </ul>
      
      <p><a href="filter.html">Filter a SingleCellAssay</a></p>
      <ul>
        <li><code><a href="filter.html">burdenOfFiltering</a></code></li><li><code><a href="filter.html">filter</a></code></li>
      </ul>
      
      <p><a href="freq.html">Summary statistics for genes in an experiment</a></p>
      <ul>
        <li><code><a href="freq.html">condSd</a></code></li><li><code><a href="freq.html">condmean</a></code></li><li><code><a href="freq.html">freq</a></code></li><li><code><a href="freq.html">numexp</a></code></li>
      </ul>
      
      <p><a href="FromFlatDF.html">Construct a SingleCellAssay (or derived subclass) from a `flat` (melted) data.frame/data.table</a></p>
      <ul>
        <li><code><a href="FromFlatDF.html">FluidigmAssay</a></code></li><li><code><a href="FromFlatDF.html">FromFlatDF</a></code></li><li><code><a href="FromFlatDF.html">SingleCellAssay</a></code></li>
      </ul>
      
      <p><a href="FromMatrix.html">Construct a SingleCellAssay from a matrix or array of expression</a></p>
      <ul>
        <li><code><a href="FromMatrix.html">FromMatrix</a></code></li>
      </ul>
      
      <p><a href="getConcordance.html">Get the concordance between two experiments</a></p>
      <ul>
        <li><code><a href="getConcordance.html">getConcordance</a></code></li><li><code><a href="getConcordance.html">getrc</a></code></li><li><code><a href="getConcordance.html">getss</a></code></li><li><code><a href="getConcordance.html">getwss</a></code></li>
      </ul>
      
      <p><a href="melt.SingleCellAssay.html">&quot;Melt&quot; a </a></p>
      <ul>
        <li><code><a href="melt.SingleCellAssay.html">melt.SingleCellAssay</a></code></li>
      </ul>
      
      <p><a href="primerAverage.html">Average expression values for duplicated/redundant genes</a></p>
      <ul>
        <li><code><a href="primerAverage.html">primerAverage</a></code></li>
      </ul>
      
      <p><a href="read.fluidigm.html">Reads a Fluidigm Biomark (c. 2011) raw data file (or set of files)</a></p>
      <ul>
        <li><code><a href="read.fluidigm.html">read.fluidigm</a></code></li>
      </ul>
      
      <p><a href="zlm.html">Zero-inflated regression for SingleCellAssay</a></p>
      <ul>
        <li><code><a href="zlm.html">zlm</a></code></li><li><code><a href="zlm.html">zlm.SingleCellAssay</a></code></li>
      </ul>
          <h2 id="section-thresholding" class="hasAnchor"><a href="#section-thresholding" class="anchor"></a>Thresholding</h2>
      <p class="section-desc"><p>Adaptive thresholding of background noise.</p></p>

      
      <p><a href="plot.thresholdSCRNACountMatrix.html">Plot cutpoints and densities for thresholding</a></p>
      <ul>
        <li><code><a href="plot.thresholdSCRNACountMatrix.html">plot.thresholdSCRNACountMatrix</a></code></li>
      </ul>
      
      <p><a href="summary.thresholdSCRNACountMatrix.html">Summarize the effect of thresholding</a></p>
      <ul>
        <li><code><a href="summary.thresholdSCRNACountMatrix.html">print.summaryThresholdSCRNA</a></code></li><li><code><a href="summary.thresholdSCRNACountMatrix.html">summary.thresholdSCRNACountMatrix</a></code></li>
      </ul>
      
      <p><a href="thresholdSCRNACountMatrix.html">Threshold a count matrix using an adaptive threshold.</a></p>
      <ul>
        <li><code><a href="thresholdSCRNACountMatrix.html">thresholdSCRNACountMatrix</a></code></li>
      </ul>
          <h2 id="section-differential-expression-testing-with-hurdle-model" class="hasAnchor"><a href="#section-differential-expression-testing-with-hurdle-model" class="anchor"></a>Differential Expression Testing with Hurdle Model</h2>
      <p class="section-desc"><p>Fit a Hurdle linear model to test for zero-inflated differential expression.</p></p>

      
      <p><a href="BayesGLMlike-class.html">Wrapper for bayesian GLM</a></p>
      <ul>
        <li><code><a href="BayesGLMlike-class.html">BayesGLMlike-class</a></code></li>
      </ul>
      
      <p><a href="defaultPrior.html">Initialize a prior to be used a prior for BayeGLMlike/BayesGLMlike2</a></p>
      <ul>
        <li><code><a href="defaultPrior.html">defaultPrior</a></code></li>
      </ul>
      
      <p><a href="ebayes.html">Estimate hyperparameters for hierarchical variance model for continuous component</a></p>
      <ul>
        <li><code><a href="ebayes.html">ebayes</a></code></li>
      </ul>
      
      <p><a href="LMERlike-class.html">Wrapper for lmer/glmer</a></p>
      <ul>
        <li><code><a href="LMERlike-class.html">LMERlike-class</a></code></li><li><code><a href="LMERlike-class.html">coef,LMERlike-method</a></code></li><li><code><a href="LMERlike-class.html">logLik,LMERlike-method</a></code></li><li><code><a href="LMERlike-class.html">update,LMERlike-method</a></code></li><li><code><a href="LMERlike-class.html">vcov,LMERlike-method</a></code></li>
      </ul>
      
      <p><a href="print.summaryZlmFit.html">Print summary of a ZlmFit</a></p>
      <ul>
        <li><code><a href="print.summaryZlmFit.html">print.summaryZlmFit</a></code></li>
      </ul>
      
      <p><a href="zlm.html">Zero-inflated regression for SingleCellAssay</a></p>
      <ul>
        <li><code><a href="zlm.html">zlm</a></code></li><li><code><a href="zlm.html">zlm.SingleCellAssay</a></code></li>
      </ul>
          <h2 id="section-specifying-hypothesis-and-contrasts-to-test" class="hasAnchor"><a href="#section-specifying-hypothesis-and-contrasts-to-test" class="anchor"></a>Specifying hypothesis and contrasts to test</h2>
      <p class="section-desc"><p>Testing linear functions of coefficients.</p></p>

      
      <p><a href="Hypothesis.html">Describe a linear model hypothesis to be tested</a></p>
      <ul>
        <li><code><a href="Hypothesis.html">CoefficientHypothesis</a></code></li><li><code><a href="Hypothesis.html">Hypothesis</a></code></li>
      </ul>
      
      <p><a href="lrTest.html">Run a likelihood-ratio test</a></p>
      <ul>
        <li><code><a href="lrTest.html">lrTest</a></code></li>
      </ul>
      
      <p><a href="waldTest.html">Run a Wald test</a></p>
      <ul>
        <li><code><a href="waldTest.html">waldTest</a></code></li>
      </ul>
          <h2 id="section-low-level-manipulation-of-fitted-hurdle-models" class="hasAnchor"><a href="#section-low-level-manipulation-of-fitted-hurdle-models" class="anchor"></a>Low-level manipulation of fitted hurdle models</h2>
      <p class="section-desc"><p>Return coefficients, standard errors, etc. Calculate and return residuals from models.</p></p>

      
      <p><a href="collectResiduals.html">Residual hooks and collection methods</a></p>
      <ul>
        <li><code><a href="collectResiduals.html">collectResiduals</a></code></li><li><code><a href="collectResiduals.html">combined_residuals_hook</a></code></li><li><code><a href="collectResiduals.html">continuous_residuals_hook</a></code></li><li><code><a href="collectResiduals.html">deviance_residuals_hook</a></code></li><li><code><a href="collectResiduals.html">discrete_residuals_hook</a></code></li><li><code><a href="collectResiduals.html">fitted_phat</a></code></li><li><code><a href="collectResiduals.html">partialScore</a></code></li>
      </ul>
      
      <p><a href="impute.html">impute missing continuous expression for plotting</a></p>
      <ul>
        <li><code><a href="impute.html">impute</a></code></li>
      </ul>
      
      <p><a href="logFC.html">Calculate log-fold changes from hurdle model components</a></p>
      <ul>
        <li><code><a href="logFC.html">getLogFC</a></code></li><li><code><a href="logFC.html">logFC</a></code></li>
      </ul>
      
      <p><a href="predict.ZlmFit.html">Return predictions from a ZlmFit object.</a></p>
      <ul>
        <li><code><a href="predict.ZlmFit.html">predict.ZlmFit</a></code></li>
      </ul>
      
      <p><a href="se.coef.html">Return coefficient standard errors</a></p>
      <ul>
        <li><code><a href="se.coef.html">se.coef</a></code></li>
      </ul>
      
      <p><a href="ZlmFit-class.html">An S4 class to hold the output of a call to zlm</a></p>
      <ul>
        <li><code><a href="ZlmFit-class.html">ZlmFit-class</a></code></li><li><code><a href="ZlmFit-class.html">coef,ZlmFit-method</a></code></li><li><code><a href="ZlmFit-class.html">lrTest,ZlmFit,CoefficientHypothesis-method</a></code></li><li><code><a href="ZlmFit-class.html">lrTest,ZlmFit,Hypothesis-method</a></code></li><li><code><a href="ZlmFit-class.html">lrTest,ZlmFit,matrix-method</a></code></li><li><code><a href="ZlmFit-class.html">se.coef,ZlmFit-method</a></code></li><li><code><a href="ZlmFit-class.html">vcov,ZlmFit-method</a></code></li><li><code><a href="ZlmFit-class.html">waldTest,ZlmFit,CoefficientHypothesis-method</a></code></li><li><code><a href="ZlmFit-class.html">waldTest,ZlmFit,Hypothesis-method</a></code></li>
      </ul>
          <h2 id="section-gene-set-enrichment-testing" class="hasAnchor"><a href="#section-gene-set-enrichment-testing" class="anchor"></a>Gene set enrichment testing</h2>
      <p class="section-desc"><p>Tests on the average differential expression effect in a gene set, accounting for gene-gene correlations. Uses bootstraps to assess the gene-gene correlations.</p></p>

      
      <p><a href="bootVcov1.html">Bootstrap a zlmfit</a></p>
      <ul>
        <li><code><a href="bootVcov1.html">bootVcov1</a></code></li>
      </ul>
      
      <p><a href="gseaAfterBoot.html">Gene set analysis for hurdle model</a></p>
      <ul>
        <li><code><a href="gseaAfterBoot.html">gseaAfterBoot</a></code></li>
      </ul>
      
      <p><a href="pbootVcov1.html">Bootstrap a zlmfit</a></p>
      <ul>
        <li><code><a href="pbootVcov1.html">pbootVcov1</a></code></li>
      </ul>
      
      <p><a href="summary-GSEATests-method.html">Summarize gene set enrichment tests</a></p>
      <ul>
        <li><code><a href="summary-GSEATests-method.html">summary,GSEATests-method</a></code></li>
      </ul>
          <h2 id="section-plotting-methods" class="hasAnchor"><a href="#section-plotting-methods" class="anchor"></a>Plotting methods</h2>
      <p class="section-desc"><p>Miscellaneous plotting methods. Concordance, PCA biplots, and effects plots for fitted hurdle models.</p></p>

      
      <p><a href="myBiplot.html">Makes a nice BiPlot</a></p>
      <ul>
        <li><code><a href="myBiplot.html">myBiplot</a></code></li>
      </ul>
      
      <p><a href="plotSCAConcordance.html">Concordance plots of filtered single vs n-cell assays</a></p>
      <ul>
        <li><code><a href="plotSCAConcordance.html">plotSCAConcordance</a></code></li>
      </ul>
      
      <p><a href="stat_ell.html">Plot confidence ellipse in 2D</a></p>
      <ul>
        <li><code><a href="stat_ell.html">stat_ell</a></code></li>
      </ul>
    
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-loading-transforming-and-summarizing">Loading, transforming and summarizing</a></li>
      <li><a href="#section-thresholding">Thresholding</a></li>
      <li><a href="#section-differential-expression-testing-with-hurdle-model">Differential Expression Testing with Hurdle Model</a></li>
      <li><a href="#section-specifying-hypothesis-and-contrasts-to-test">Specifying hypothesis and contrasts to test</a></li>
      <li><a href="#section-low-level-manipulation-of-fitted-hurdle-models">Low-level manipulation of fitted hurdle models</a></li>
      <li><a href="#section-gene-set-enrichment-testing">Gene set enrichment testing</a></li>
      <li><a href="#section-plotting-methods">Plotting methods</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Andrew McDavid, Greg Finak, Masanao Yajima.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
