##' Methods for analysing single cell assay data
##'
##' This packages provides data structures and functions for statistical analysis of single-cell assay data such as Fluidigm single cell gene expression assays.
##'
##' @name SingleCellAssay-package
##' @aliases SingleCellAssay-package
##' @docType package
##' @title "Tools for Single Cell Assay Analysis"
##' @keywords package
##' @rdname SingleCellAssay-package
##' @seealso \code{\link{SCASet}}
NULL
###=========GENERICS==================

##' Accessor for wellKey
##'
##' This returns the wellKey, which is a unique identifier generated by \code{idvars} in the \code{mapping}
##' @title getwellKey
##' @param sc An object with a \code{wellKey}
##' @return \code{integer} giving the unique id generated
##' @name getwellKey-method
##' 
##' @exportMethod getwellKey
##' @docType methods
##' @rdname getwellKey-methods
##' @keywords accessor
setGeneric('getwellKey', function(sc) standardGeneric('getwellKey'))



##' Accessor for cellData \code{AnnotatedDataFrame}
##'
##' Returns the \code{cellData}.
##' @title cData
##' @param sc An object with \code{cellData}
##' @return \code{AnnotatedDataFrame}
##'
##' 
##' @export
##' @docType methods
##' @rdname cellData-methods
##' @keywords accessor
setGeneric('cellData', function(sc) standardGeneric('cellData'))

##' Accessor for featureData \code{data.frame}
##'
##' Returns the \code{featureData} \code{data.frame}.
##' @title fData
##' @param object An object with \code{featureData}
##' @return \code{data.frame} 
##' @docType methods
##' @rdname fData-methods
##' @keywords accessor
##' @name fData
##' @aliases fData,SingleCellAssay-method
##' @importMethodsFrom Biobase fData
#setGeneric('fData', function(sc) standardGeneric('fData'))
NULL

##' Accessor for featureData \code{AnnotatedDataFrame}
##'
##' Returns the \code{featureData}.
##' @param object An object with \code{featureData}
##' @return \code{AnnotatedDataFrame}
##' @docType methods
##' @rdname featureData-methods
##' @keywords accessor
##' @name featureData
##' @aliases featureData,SingleCellAssay-method
##' @importMethodsFrom Biobase featureData
#setGeneric('featureData', function(object) standardGeneric('featureData'))
NULL

##' Melt a rectangular array
##'
##' Return a 'molten' (flat) representation of a rectangular array
##'
##' @param data A rectangular array, with attributes attached to its rows and
##' columns
##' @param ... ignored
##' @return A \code{data.frame} typically, with the cartesian product of the
##' row and column attributes and the values from the rectangular array
##' 
##' 
##' @rdname melt
##' @title melt
##' @aliases melt
##' @keywords transformation
##' @importFrom reshape melt
##' @S3method melt SingleCellAssay
## melt.SingleCellAssay<-function(data,...){
##   m <- melt.data.frame(cbind(cData(data), exprs(data)), id.vars=names(cData(data)), variable_name='primerid')
##   m <- merge(m, fData(data), by='primerid')
##   if(data@keep.names) return(rename(m, c('value'=dimnames(data)[[3]][layer(data)])))
##   return(m)
## }
melt.SingleCellAssay<-function(data,...){
  m <- cbind(
    cData(data)[rep(seq_len(nrow(data)), ncol(data)),],
    fData(data)[rep(seq_len(ncol(data)), each=nrow(data)),],
    value=as.vector(exprs(data)))
  rn <-  c('value'=dimnames(data)[[3]][SingleCellAssay:::layer(data)])
  if(data@keep.names) return(rename(m,rn))
  return(m)
}




mkunique<-function(x,G){
    cbind(x,primerid.unk=make.unique(as.character(get(G,x))))
     }


check.vars <- function(cellvars, featurevars, phenovars, dataframe, nc, nr){
  if( any(cellvars %in% featurevars))
    stop("'cellvars', 'idvars' must be disjoint from 'featurevars', 'primerid', 'geneid'")
  cvars.in <- cellvars %in% names(dataframe)
  fvars.in <- featurevars %in% names(dataframe)
  if( !all(cvars.in)) stop(cellvars[!cvars.in][1], ' not found')
  if( !all(fvars.in)) stop(featurevars[!fvars.in][1], ' not found')
  
  nuniquef<-nrow(uniqueModNA(dataframe[,featurevars, drop=FALSE], 'primerid'))
  nuniquec<-nrow(uniqueModNA(dataframe[,cellvars, drop=FALSE], 'wellKey'))  
  if(nuniquef != nc)
    stop("'featurevars' must be keyed by 'primerid'")
  if(nuniquec != nr)
    stop("'cellvars' must be keyed by 'idvars'")
}

## might have bad complexity, but could construct one at time, then glue cheaply
## Not too bad except for deduplication.. will use data.table
##' @import data.table
fixdf <- function(df, idvars, primerid, measurement, cmap, fmap, keep.names){
  if(!inherits(df,"data.frame")){
    stop("Argument `dataframe` should be a data.frame.")
  }
  if(!all(idvars %in% colnames(df))){
    stop("Invalid idvars column name. Not in data.frame")
  }
  if(!all(primerid%in%colnames(df))){
    stop("Invalid primerid column name. Not in data.frame")
  }
  if(!all(measurement%in%colnames(df))){
    stop("Invalid measurement column name. Not in data.frame")
  }
  ## FIXME: should check if cmap and fmap are in df and throw intelligible error

  bothMap <- c(cmap, fmap)
  for(nm in names(bothMap)){
    if(nm %in% colnames(df) && bothMap[[nm]] != nm){
      names(df) <- make.unique(c(nm, colnames(df)))[-1]
    warning("renaming column ", nm)
    }
      if(!all(bothMap[[nm]] %in% names(df))){
        stop('could not find column named ', bothMap[[nm]], ' in dataframe')
      }
    rn <- nm
    names(rn) <- as.character(bothMap[[nm]])
    if(keep.names){ df[,rn] <- df[,names(rn)]}
    else{ df <- rename(df, rn)}
  }

  wk <- do.call(paste, df[,idvars, drop=FALSE])
  cellCounts <- table(wk)
  pid <- do.call(paste, df[,primerid, drop=FALSE])
  primerCounts <- table(pid)
  df[,'wellKey'] <- wk
  df[,'primerid'] <- pid
  incomplete <- !all(cellCounts == cellCounts[1])
  dupPrimers <- table(df$wellKey, df$primerid, exclude=NULL) #cross tab of primerid x wellKey
  duped.primers <- apply(dupPrimers>1, 2, which)
  duped.primers <- duped.primers[sapply(duped.primers, length)>0]
  if(length(duped.primers)>0){
    warning("Primerid ", names(duped.primers)[1], " appears be duplicated.\n I will attempt to make it unique, but this may fail if the order of the primers is inconsistent in the dataframe.")
    dt<-data.table(df)
    #dt$primer.orig <- dt$primerid
    dt[,primerid.orig:=primerid]
    dt[,primerid:=make.unique(.SD$primerid.orig),by='wellKey']
    df<-as.data.frame(dt)
#    df <- ddply(df,'wellKey',mkunique,G='primerid')
#    df[,'primerid.orig'] <- df[,'primerid']
#    df[,'primerid'] <- df[,'primerid.unk']
#    df[,'primerid.unk'] <- NULL
    fmap['primerid.orig'] <- 'primerid.orig'
  }
  
  if(incomplete){
    message("dataframe appears incomplete, attempting to complete it with NAs")
    skeleton <- expand.grid.df(unique(df[,"primerid", drop=FALSE]), unique(df[, "wellKey", drop=FALSE]))
    df <- merge(skeleton, df, all.x=TRUE, by=c('primerid', 'wellKey'))
  }

  ord <- do.call(order, df[, c("primerid", "wellKey")])
  df <- df[ord,]
  list(df=df, rn=unique(df$wellKey), cn=unique(df$primerid), fmap=fmap, cmap=cmap)
}

##' @importFrom plyr ddply
##' @importFrom reshape expand.grid.df
##' @importFrom reshape rename
## unnamed arguments get passed along to callNextMethod
## which eventually just sets the slots
setMethod('initialize', 'SingleCellAssay',
          function(.Object, dataframe, idvars, primerid, measurement, cellvars=NULL, featurevars=NULL, phenovars=NULL, sort=TRUE, ...){
            ##message(class(.Object), ' calling SingleCellAssay Initialize')  #DEBUG
            .Object <- callNextMethod()
            if(sort) .Object <- sort(.Object)
            if(!missing(dataframe)){              #called using melted dataframe
              ##message('...with dataframe') #DEBUG
              if(missing(idvars) || missing(primerid) || missing(measurement)){
                stop("Must supply all of 'idvars', 'primerid' and 'measurement' if 'dataframe' is passed")
              }
              if(nrow(.Object) > 0 || ncol(.Object)>0) warning('slots will be overwritten when dataframe is provided')
              ## fixdf: make primerid unique, generate idvar column, rename columns according to cmap and fmap, complete df
              fixed <- fixdf(dataframe, idvars, primerid, measurement, .Object@cmap, .Object@fmap, .Object@keep.names)
              .Object@fmap <- fixed$fmap        #needed if we deduplicated primerid
              .Object@cmap <- fixed$cmap
              dl <- array(fixed$df[,measurement],
                          dim=c(length(fixed$rn), length(fixed$cn), length(measurement)),
                          dimnames=list(wellKey=fixed$rn, primerid=fixed$cn, layer=measurement))
              .Object@.Data <- dl
              cellvars <- union(cellvars, c('wellKey', idvars, phenovars, names(.Object@cmap))) #fixme when we support phenovars
              featurevars <- union(c('primerid', primerid, featurevars), names(.Object@fmap))
              if(.Object@keep.names){
                featurevars <- union(featurevars, unlist(.Object@fmap))
                cellvars <- union(cellvars, unlist(.Object@cmap))
              }
              check.vars(cellvars, featurevars, phenovars, fixed$df, length(fixed$cn), length(fixed$rn))
              cell.adf  <- new("AnnotatedDataFrame")
              pData(cell.adf)<-uniqueModNA(fixed$df[,cellvars, drop=FALSE], 'wellKey') #automatically sorted by idvars
              sampleNames(cell.adf) <- unique(fixed$df$wellKey)
              ##pheno.adf <- new('AnnotatedDataFrame')
              ##need a phenokey into the melted data frame for this to make sense
              f.adf <- new('AnnotatedDataFrame')
              pData(f.adf) <- uniqueModNA(fixed$df[,featurevars, drop=FALSE], 'primerid')
              sampleNames(f.adf) <- unique(fixed$df$primerid)

              ## currently take care of this in fixdf
              ## if(.Object@keep.names){
              ##   pData(cell.adf)[,unlist(.Object@cmap)] <- pData(cell.adf)[,names(.Object@cmap)]
              ##   pData(f.adf)[,unlist(.Object@fmap)] <- pData(f.adf)[,names(.Object@fmap)]
              ##   pData(f.adf)[,primerid] <- pData(f.adf)[,'primerid']
              ## }
  
              .Object@cellData<-cell.adf
              .Object@featureData <- f.adf
            }
            
            .Object
          })

sort.SingleCellAssay <- function(x, decreasing=FALSE, ...){
  if(nrow(cData(x))>0){
    rowOrd <- order(cData(x)$wellKey)
    x@.Data <- as(x, 'DataLayer')[rowOrd,]
    x@cellData <- x@cellData[rowOrd,]
  }
  if(nrow(fData(x))>0){
     colOrd <- order(fData(x)$primerid)
    x@.Data <- as(x, 'DataLayer')[,colOrd]
    x@featureData <- x@featureData[colOrd,]
  }
  stopifnot(validObject(x))
  x
}

setMethod('sort', signature=c(x='SingleCellAssay'), sort.SingleCellAssay)

uniqueModNA <- function(df, exclude){
  #browser()
  w.include <- names(df)
  if(ncol(df)>1){
  w.include <- setdiff(w.include, exclude)
}
  u <- unique(df)
  anyNa <- apply(is.na(u)[,w.include, drop=FALSE], 1, all)
  u[!anyNa,,drop=FALSE]
}

#setGeneric("melt",function(data,...){
#standardGeneric("melt")
#  UseMethod(generic="melt",data)
#  },useAsDefault=reshape::melt)

##' Evaluate an expression (returning a \code{logical} vector in the context of
##' \code{x}
##'
##' Really just syntactic sugar.
##'
##' @param x A matrix or \code{data.frame} typically
##' @param thesubset An expression, which evaluated in either the current
##' environment, or the parent, yields a logical vector
##'
##' @return The same class as \code{x} typically, with entries in which thesubset was \code{TRUE}
##' 
##' @export
##' @docType methods
##' @rdname subset-methods
##' @keywords transformation
setGeneric('subset')


##' Return a deep copy of an object
##'
##' Because SingleCellAssay keeps the datastore in an environment, it's not sufficient to create a new object to make a copy on an object
##' @param sc object
##' @return copy of sc
##' @export
##' @docType methods
##' @rdname copy-methods
setGeneric('copy', function(object) standardGeneric('copy'))


##' @rdname getwellKey-methods
##' @aliases getwellKey,SingleCellAssay-method
setMethod('getwellKey', 'SingleCellAssay', function(sc) {cData(sc)$wellKey})



##' @rdname cData-methods
##' @aliases cData,SingleCellAssay-method
##' @section Methods:
##' \describe{
##' \item{\code{sc = "SingleCellAssay"}}{}
##' }
setMethod('cData', 'SingleCellAssay', function(sc)  pData(sc@cellData))

##' @rdname cellData-methods
##' @aliases cellData,SingleCellAssay-method
setMethod('cellData', 'SingleCellAssay', function(sc) sc@cellData)


##' @rdname fData-methods
##' @aliases fData,SingleCellAssay-method
##' @exportMethod fData
setMethod('fData', 'SingleCellAssay', function(object) pData(object@featureData))

##' @rdname featureData-methods
##' @aliases featureData,SingleCellAssay-method
##' @exportMethod featureData
setMethod('featureData', 'SingleCellAssay', function(object)  object@featureData)

##' @rdname melt
##' @details \code{signature(data="SingleCellAssay")}: return a \code{data.frame}, which contains a melted version of \code{data}.
##' @aliases melt,SingleCellAssay-method
##' @exportMethod melt
setMethod("melt","SingleCellAssay",melt.SingleCellAssay )


.scaSubset <- function(x, i, j, ..., drop=FALSE){
  if(missing(i)){
    i<-1:nrow(x)
  }
  if(any(is.na(i))) stop("NAs not permitted in 'i' index")
  
  if(inherits(i,"character")){
    wk<-getwellKey(x)
    if(length(setdiff(i, wk))>0) stop('wellKeys \n', paste(setdiff(i, wk), sep=','), '\n not found!')
    i <- match(i, wk)
  }
  
  if(!missing(j)){
    if(any(is.na(j))) stop("NAs not permitted in 'j' index")  
    pk<-fData(x)$primerid
    
    if(inherits(j,"character")){
      J <- match(j, pk)
      if(!(all(j%in%pk))){
        stop("feature names \n",paste(j[!j%in%as.matrix(pk)],collapse=" "), "\n not found!");
      }
      j<-J
    }
    newfdf <- featureData(x)[j,] 
  }else {                               #j missing
    j <- TRUE
  }

  
  newcdf <- cellData(x)[i,]
  if(!exists("newfdf")){
    newfdf<-x@featureData
  }
  ## Not callNextMethod, because we want to dispatch [ not [[
  ## And we don't want to wrap this method 
  .Data <- selectMethod('[', signature='DataLayer')(x, i, j, ..., drop=drop)
  x@featureData <- newfdf
  x@cellData <- newcdf
  x@.Data <- .Data
  x
}


setMethod('[[', signature(x="SingleCellAssay"), .scaSubset)

##' @name [
##' @title subset methods
##' @details \code{signature(x="SingleCellAssay", i="ANY")}: \code{x[i]}, where \code{i} is a logical, integer, or character vector, recycled as necessary to match \code{nrow(x)}. Optional \code{x[[i,j]]} where j is a logical, integer or character vector selecting the features based on ``primerid'' which is unique, while ``geneid'' or gene name is not necessarily unique. 
##' @aliases [,SingleCellAssay,ANY-method
##' @aliases [,DataLayer,ANY-method
##' @keywords transform
##' @rdname angleBracket-methods
##' @export
setMethod("[", signature(x="SingleCellAssay"), .scaSubset)
  

##' @rdname subset-methods
##' @aliases subset,SingleCellAssay-method
##' @details \code{signature(x='SingleCellAssay', thesubset='ANY')}: Return a new SingleCellAssay consisting of cells in which thesubset is TRUE
setMethod('subset', 'SingleCellAssay', function(x, thesubset, ...){
  e <- substitute(thesubset)
  asBool <- try(eval(e, cData(x), parent.frame(n=2)), silent=TRUE)
  if(inherits(asBool, 'try-error')) stop(paste('Variable in subset not found:', strsplit(asBool, ':')[[1]][2]))
  #this is a special case of "subset", not of the "[[" method, so..
  if(length(asBool)==1){
    if(asBool==TRUE){
      x 
    }else{
      x[[asBool]]
    }
  }else{
    x[[asBool]]
  }
})

##' Split into SCASet
##'
##' Splits a \code{SingleCellAssay} into a \code{SCASet} by a factor (or something coercible into a factor) or a character giving a column of the melted SingleCellAssay
##' @param x SingleCellAssay
##' @param f length-1 character or factor of length nrow(x)
##' @return SCASet
##' @docType methods
##' @rdname split-methods
##' @aliases split,SingleCellAssay,ANY-method
##' @aliases split,DataLayer,ANY-method
##' @name split
##' @exportMethod split
setMethod('split', signature(x='SingleCellAssay'), 
          function(x, f, drop=FALSE, ...){
  ## Split a SingleCellAssay by criteria
  ###f must be a character naming a cData variable
  if(inherits(f, 'character')){
    if(length(f) != nrow(x)){
      f <- lapply(cData(x)[,f, drop=FALSE], as.factor)
    } else{
      f <- list(as.factor(f))
    }
  } else if(inherits(f, 'factor')){
    f <- list(f)
  }
     all.factor <- all(sapply(f, is.factor))
     if(!all.factor) stop('f must be character vector naming columns of x; or factor; or list of factors')
     all.length <- all(sapply(f, length)==nrow(x))
     if(!all.length) stop('each element in f must be length nrow(x)')
  out <- callNextMethod()
  cD <- split.data.frame(x@cellData, f)
  for(i in seq_along(out)){
    out[[i]]@cellData <- cD[[i]]
    out[[i]]@id <- names(out)[i]
  }
  new('SCASet', set=out)
})

.SingleCellAssayCombine <- function(scalist){
  names(scalist)[1:2] <- c('x', 'y')
  do.call(combine, scalist)
}


## FIXME: gdata (not sure why it's imported) shadows the generic definition
##'Combine two SingleCellAssay or derived classes
##'
##' combines two single cell assays provided they share a common mapping
##' TODO combining based on a common mapping may be too restrictive. This may change depending on needs.
##' @importMethodsFrom BiocGenerics combine
##' @exportMethod combine
##' @aliases combine,SingleCellAssay,SingleCellAssay-method
##' @docType methods
##' @rdname combine-methods
setMethod('combine', signature(x='SingleCellAssay', y='SingleCellAssay'), function(x, y, ...) {
  proto <- callNextMethod()
  cellData <- combine(cellData(x), cellData(y))
  ## Combine wants to do a rbind-like operation with ADFs
  ## But we want a Cbind
  nx <- names(fData(x))
  ny <- names(fData(y))
  featureData <- featureData(x)
  pData(featureData) <- cbind(fData(x), fData(y)[,setdiff(ny, nx)])
  proto@cellData <- cellData
  proto@featureData <- featureData
  proto
})


##' @export
getMapping <- function(x, map){
  return(list(map))
}


##' @exportMethod copy
##' @aliases copy,SingleCellAssay-method
##' @rdname copy-methods
setMethod('copy', 'SingleCellAssay',
          function(object){
            o2 <- object[[1:nrow(object)]]
            o2
          })


setMethod('show', 'SingleCellAssay', function(object){
callNextMethod()
cat(' id: ', object@id, '\n')
})

setMethod('combine', signature=c(x='SingleCellAssay', y='ANY'), function(x, y, ...){
  adf <- new('AnnotatedDataFrame')
  df <- data.frame(y)
  names(df) <- deparse(substitute(y))
  pData(adf) <- df
  selectMethod('combine', c(x=class(x), y='AnnotatedDataFrame'))(x, adf)
})

setMethod('combine', signature=c(x='SingleCellAssay', y='data.frame'), function(x, y, ...){
  adf <- new('AnnotatedDataFrame')
  pData(adf) <- y
  selectMethod('combine', c(x=class(x), y='AnnotatedDataFrame'))(x, adf)
})

setMethod('combine', signature=c(x='SingleCellAssay', y='AnnotatedDataFrame'), function(x, y, ...){
    if(nrow(x) == ncol(x)) stop("x has same number of rows and columns, must explicitly specify 'along'")
    if(nrow(y) == nrow(x)) along <- 'cellData'
    else if(nrow(y) == ncol(x)) along <- 'featureData'
    else stop('Dimension mismatch between y and x')
    
  if(length(intersect(along , c('cellData', 'featureData')))!=1) stop("If specified, along must be either 'cellData', or 'featureData'")
  newdata <- slot(x, along)
  pData(newdata) <- cbind(pData(newdata), pData(y))
  slot(x, along) <- newdata
  x
})
